File: src/frontend/src/api/projects/state.ts
import { apiJson, apiRequest } from "../client";
import type { CredentialProfile } from "./types";

export async function listCredentialProfiles(): Promise<CredentialProfile[]> {
  const res = await apiRequest("/api/state/credential-profiles", {
    credentials: "include",
  });
  const data = await apiJson<{ profiles: CredentialProfile[] }>(res);
  return data.profiles;
}

export async function createCredentialProfile(payload: {
  name: string;
  provider: string;
  credentials: Record<string, unknown>;
  meta?: Record<string, unknown>;
}): Promise<CredentialProfile> {
  const res = await apiRequest("/api/state/credential-profiles", {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return apiJson<CredentialProfile>(res);
}

export async function updateCredentialProfile(
  profileId: string,
  payload: {
    name?: string;
    credentials?: Record<string, unknown>;
    meta?: Record<string, unknown>;
  },
): Promise<CredentialProfile> {
  const res = await apiRequest(`/api/state/credential-profiles/${profileId}`, {
    method: "PUT",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return apiJson<CredentialProfile>(res);
}

export async function deleteCredentialProfile(profileId: string): Promise<void> {
  const res = await apiRequest(`/api/state/credential-profiles/${profileId}`, {
    method: "DELETE",
    credentials: "include",
  });
  await apiJson<{ ok: boolean }>(res);
}

File: src/backend/app/routers/state.py
"""User-scoped state backend utility endpoints."""
from __future__ import annotations

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from app.core.config import get_settings
from app.models import User
from app.routers import auth_dependencies as auth_deps
from app.routers.http_errors import raise_http_error
from app.services.state_backends import credential_profiles

router = APIRouter(prefix="/api/state", tags=["state"])


class CredentialProfileCreateBody(BaseModel):
    name: str
    provider: str
    credentials: dict = Field(default_factory=dict)
    meta: dict = Field(default_factory=dict)


class CredentialProfileUpdateBody(BaseModel):
    name: str | None = None
    credentials: dict | None = None
    meta: dict | None = None


@router.get("/credential-profiles")
async def list_profiles(user: User = Depends(auth_deps.require_current_user)) -> dict:
    settings = get_settings()
    profiles = await credential_profiles.list_credential_profiles(
        user_id=user.id,
        secret=settings.state_encryption_key,
    )
    return {"profiles": profiles}


@router.post("/credential-profiles")
async def create_profile(
    body: CredentialProfileCreateBody,
    user: User = Depends(auth_deps.require_current_user),
) -> dict:
    settings = get_settings()
    try:
        profile = await credential_profiles.create_credential_profile(
            user_id=user.id,
            name=body.name,
            provider=body.provider,
            credentials=body.credentials,
            meta=body.meta,
            secret=settings.state_encryption_key,
        )
        return profile
    except ValueError as exc:
        raise_http_error(400, code=str(exc), message=str(exc))


@router.put("/credential-profiles/{profile_id}")
async def update_profile(
    profile_id: str,
    body: CredentialProfileUpdateBody,
    user: User = Depends(auth_deps.require_current_user),
) -> dict:
    settings = get_settings()
    try:
        profile = await credential_profiles.update_credential_profile(
            user_id=user.id,
            profile_id=profile_id,
            name=body.name,
            credentials=body.credentials,
            meta=body.meta,
            secret=settings.state_encryption_key,
        )
        return profile
    except ValueError as exc:
        status = 404 if str(exc) == "profile_not_found" else 400
        raise_http_error(status, code=str(exc), message=str(exc))


@router.delete("/credential-profiles/{profile_id}")
async def delete_profile(
    profile_id: str,
    user: User = Depends(auth_deps.require_current_user),
) -> dict:
    deleted = await credential_profiles.delete_credential_profile(user_id=user.id, profile_id=profile_id)
    if not deleted:
        raise_http_error(404, code="profile_not_found", message="Credential profile not found")
    return {"ok": True}
